### PR TITLE
fix invalidations

### LINKED ISSUE #12: Unrolled.jl invalidates quite a lot
Thirst of all, thanks for the nice package! I get the following result with Julia v1.8:

```julia
julia> using Pkg; Pkg.activate(temp=true); Pkg.add("Unrolled")
  Activating new project at `/tmp/jl_n64URx`
    Updating registry at `~/.julia/registries/General.toml`
   Resolving package versions...
    Updating `/tmp/jl_n64URx/Project.toml`
  [9602ed7d] + Unrolled v0.1.3
    Updating `/tmp/jl_n64URx/Manifest.toml`
  [1914dd2f] + MacroTools v0.5.9
  [9602ed7d] + Unrolled v0.1.3
  [2a0f44e3] + Base64
  [b77e0a4c] + InteractiveUtils
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [9a3f8284] + Random
  [ea8e919c] + SHA v0.7.0
  [9e88b42a] + Serialization
  [8dfed614] + Test

julia> using SnoopCompileCore; invalidations = @snoopr(using Unrolled); using SnoopCompile

julia> trees = invalidation_trees(invalidations);

julia> length(uinvalidated(invalidations))
795

julia> trees
1-element Vector{SnoopCompile.MethodInvalidations}:
 inserting getindex(seq, ::FixedRange{A, B}) where {A, B} in Unrolled at /home/hendrik/.julia/packages/Unrolled/26uDc/src/range.jl:23 invalidated:
   mt_backedges: 1: signature Tuple{typeof(getindex), Base.Broadcast.Broadcasted, Any} triggered MethodInstance for Base.Broadcast.copyto_nonleaf!(::Array, ::Base.Broadcast.Broadcasted, ::Base.OneTo, ::Int64, ::Int64) (0 children)
                 2: signature Tuple{typeof(getindex), Base.Broadcast.Broadcasted, Any} triggered MethodInstance for Base.Broadcast.copyto_nonleaf!(::BitArray, ::Base.Broadcast.Broadcasted, ::Base.OneTo, ::Int64, ::Int64) (0 children)
                 3: signature Tuple{typeof(getindex), Base.Broadcast.Broadcasted, Any} triggered MethodInstance for Base.Broadcast.copyto_nonleaf!(::Any, ::Base.Broadcast.Broadcasted, ::Base.OneTo, ::Any, ::Int64) (93 children)
   backedges: 1: superseding getindex(wkh::WeakKeyDict{K}, key) where K in Base at weakkeydict.jl:185 with MethodInstance for getindex(::WeakKeyDict{Any, Any}, ::Any) (1 children)
              2: superseding getindex(::Type{Any}, vals...) in Base at array.jl:419 with MethodInstance for getindex(::Type{Any}, ::Vararg{Any}) (2 children)
              3: superseding getindex(h::Dict{K, V}, key) where {K, V} in Base at dict.jl:496 with MethodInstance for getindex(::Dict{String, Set{String}}, ::Any) (10 children)
              4: superseding getindex(::Type{T}, vals...) where T in Base at array.jl:403 with MethodInstance for getindex(::Type{String}, ::Any) (51 children)
              5: superseding getindex(d::IdDict{K, V}, key) where {K, V} in Base at iddict.jl:106 with MethodInstance for getindex(::IdDict{Any, Any}, ::Any) (175 children)
              6: superseding getindex(::Type{Any}, vals...) in Base at array.jl:419 with MethodInstance for getindex(::Type{Any}, ::Any) (903 children)
   7 mt_cache
```

It would be great to fix these invalidations. However, I am not sure what the best approach would be. Any suggestions?

### PR DESCRIPTION
**Pull request by ranocha**
See discussion in #12. I restricted `getindex` with `FixedRange` to tuples and abstract arrays. This removes all invalidations from Unrolled.jl :partying_face: 

```julia
julia> using SnoopCompileCore; invalidations = @snoopr(using Unrolled); using SnoopCompile

julia> trees = invalidation_trees(invalidations);

julia> length(uinvalidated(invalidations))
0

(Unrolled) pkg> st
Project Unrolled v0.0.0
Status `~/.julia/dev/Unrolled/Project.toml`
  [1914dd2f] MacroTools v0.5.9
```

Closes #12

### CODE REVIEW

File: src/range.jl [hunk 1]
```diff
@@ -20,7 +20,10 @@ Base.lastindex(fr::FixedRange) = length(fr)
 replace_end(::FixedEnd{N}, ::Type{SEQ}) where {N, SEQ} = type_length(SEQ) - N
 replace_end(n::Int, ::Type) = n
 
-@generated Base.getindex(seq, ::FixedRange{A, B}) where {A, B} =
+Base.getindex(seq::NTuple{N, Any}, fr::FixedRange{A, B}) where {N, A, B} = _getindex(seq, fr)
```
Comment (cstjean) [inline src/range.jl:23]: Why don't you type it here too? The result will be an `NTuple`.
```suggestion
Base.getindex(seq::NTuple{N, T}, fr::FixedRange{A, B}) where {N, A, B, T} = _getindex(seq, fr)::NTuple{B-A+1, T}
```

**Review by cstjean — COMMENTED** [review]
👍

File: src/range.jl (hunk 1 above)
Comment (ranocha) [reply src/range.jl]: Thanks; done in f2407af

**Comment by cstjean** [comment]
If you tested locally the change I suggested (or you want to rebase on master to have github CI), then this is good to merge!

**Comment by cstjean** [comment]
Oh, wait, should we add tests for these invalidations? 🤔 I haven't dealt with SnoopCompile yet, what's the standard thing to do here?

I'm fine with merging as is if the solution is painful, though.

**Comment by ranocha** [comment]
Let me have a look.

**Comment by ranocha** [comment]
It should be fine to merge this as it is. I'm looking for another CI script to check for invalidations.

**Comment by ranocha** [comment]
The tests fail with the additional type assertions since we can't subtract, say, `FixedEnd{1}` and `Int64`s. But the invalidations are also gone without these assertions, so it should be fine to merge this as it is (and make a new release).

**Comment by cstjean** [comment]
Thanks!

**Pull request merged**